server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java

```diff
@@ -161,15 +165,63 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         }
 
         static ClusterState createDataStream(ClusterState currentState, Request request) {
+            List<String> validationErrors = new ArrayList<>();
             if (currentState.metaData().dataStreams().containsKey(request.name)) {
-                throw new IllegalArgumentException("data_stream [" + request.name + "] already exists");
+                validationErrors.add("data_stream [" + request.name + "] already exists");
             }
 
-            MetaData.Builder builder = MetaData.builder(currentState.metaData()).put(
-                new DataStream(request.name, request.timestampFieldName, Collections.emptyList()));
+            validationErrors.addAll(validateDataStreamName(request.name));
 
-            logger.info("adding data stream [{}]", request.name);
-            return ClusterState.builder(currentState).metaData(builder).build();
+            if (currentState.metaData().hasIndex(request.name)) {
+                validationErrors.add("data_stream [" + request.name + "] conflicts with existing index");
+            }
+
+            if (currentState.metaData().hasAlias(request.name)) {
+                validationErrors.add("data_stream [" + request.name + "] conflicts with existing alias");
+            }
+
+            final String backingIndexPrefix = (request.name.startsWith(".") ? "" : ".") + request.name + "-";
+            for (String indexName : currentState.metaData().getConcreteAllIndices()) {
+                if (indexName.startsWith(backingIndexPrefix)) {
+                    validationErrors.add(
+                        "data_stream [" + request.name + "] could create backing indices that conflict with existing indices");
+                    break;
+                }
+            }
+
+            if (validationErrors.isEmpty()) {
+                MetaData.Builder builder = MetaData.builder(currentState.metaData()).put(
+                    new DataStream(request.name, request.timestampFieldName, Collections.emptyList()));
+                logger.info("adding data stream [{}]", request.name);
+                return ClusterState.builder(currentState).metaData(builder).build();
+            } else {
+                ValidationException ex = new ValidationException();
+                ex.addValidationErrors(validationErrors);
+                throw new IllegalArgumentException(ex);
+            }
+        }
+
+        private static List<String> validateDataStreamName(String name) {
```

Comment:
I think `MetaDataCreateIndexService.validateIndexName()` or `MetaDataCreateIndexService.validateIndexOrAliasName()` could be reused?